### PR TITLE
Use hatch and tbump

### PR DESCRIPTION
**Pull request by blink1073**
Hatch is newly added pypa backend with more flexibility, and helps avoid https://github.com/pypa/pip/issues/11110, which is breaking builds.

Tbump is useful for keeping versions in sync across files.  We use in in Jupyter Releaser.

### CODE REVIEW

**Review by kevin-bates — COMMENTED** [review]
Hey @blink1073 - thanks for opening this PR and ensuring the entire organization (particularly Jupyter Server) is all on the same page.  The `pip` issue with `flit` was bad luck and it sounds like `hatch` is where we, as an organization, want to be. `tbump` strikes me as pretty handy as well.

I've left a few comments, mostly due to my naivete in this area.  Again, thank you for doing this!

File: pyproject.toml
Comment (blink1073) [reply pyproject.toml:2]: No, it complicates the package for downstream users.  There actually is a way for `hatch` plugins to provide dependencies for editable mode only, which means that `hatch_jupyter_builder` could install the git hook for us.  I think it makes sense for `tbump` and `pre-commit` to be in the `test` requirements.   Ideally we could have something like `dev` requirements that extends test requirements, but you can't do that with `pyproject.toml`.

File: pyproject.toml
Comment (blink1073) [reply pyproject.toml:7]: Yeah, `tbump` is managing them all for us, it's source of truth is `[tool.tbump.version]`.

File: release.sh
Comment (blink1073) [reply release.sh:197]: That makes sense to me

Comment (kevin-bates) [reply release.sh:197]: FWIW, I applied these changes in [this commit](https://github.com/jupyter-server/enterprise_gateway/commit/11ecf18e39c62b5fb3b8fa2bcd14dcca13b7fade).  I can merge onto this PR or add in later but would appreciate your review in either case.

Also, should the `pyproject.toml`'s `[build-system][requires]` list be included in `list-build-deps.py` so that `hatchling` gets installed?  Seems like it should.

Comment (kevin-bates) [reply release.sh:197]: Would you happen to know what this is about? https://github.com/kevin-bates/enterprise_gateway/runs/6480277396?check_suite_focus=true#step:12:40

Comment (blink1073) [reply release.sh:197]: Actually, maybe a better option is to have `dev` and `test` extras, and use `pip install -e "[dev,test]"` in the docs and the Makefile?  We shouldn't need `hatchling` outside of when it gets invoked by `build` or `pip`, so I don't think we need it in `dev` dependencies.

Tbump doesn't like you to have any uncommitted changes.  What is changing those two files during the run?

Comment (kevin-bates) [reply release.sh:197]: > Actually, maybe a better option is to have dev and test extras, and use pip install -e "[dev,test]" in the docs and the Makefile? 

I'm good with that. Just not aware of the options.  Is this something you were going to do, or should I do this (although still a little unsure of pyproject.toml section names, etc.)?

> Tbump doesn't like you to have any uncommitted changes. What is changing those two files during the run?

I don't think `pre-commit` was run, so I suspect that's what's going on.  There's nothing I've introduced to explicitly change these files.

Tell you what.  Since the build is green with _this_ PR, I'm inclined to merge this PR, then follow up with some helper targets, etc. - unless you'd like to make those changes here.  At his point today I'm going to wait until tomorrow for the merge anyway, so we can talk further tomorrow if necessary.  Have a good evening!

Comment (blink1073) [reply release.sh:197]: I'll probably do a sweep of all repos to do the "pip install -e ".[dev,test]"` thingy.

> Have a good evening!

You too!

Comment (kevin-bates) [reply release.sh:197]: ok - sounds good.  Let's proceed with this PR as-is and we can circle back to add a helper target using the editable install later.

**Review by kevin-bates — APPROVED** [review]
LGTM

**Pull request merged**
